src/atheros/atherosd.c

```diff
@@ -413,51 +351,45 @@ void *serial_monitor(void *arg)
     uint8_t messageBuffer[64];
     uint8_t messageBufferIndex = 0;
 
-    uint8_t magicBytes[] = {0x47, 0x41, 0x4e, 0x53};
     uint8_t messageLengthArray[] = {9, 6};
 
     while(1)
     {
-        pthread_mutex_lock(&state->atmegafp_mutex);
-        int data = getc(state->atmegafp_read);
-        pthread_mutex_unlock(&state->atmegafp_mutex);
+        int8_t data = fgetc(state->atmegafp);
 
-        if(data != EOF)
-        {
-            messageBuffer[messageBufferIndex] = data;
+        messageBuffer[messageBufferIndex] = data;
 
-            if(messageBufferIndex <= 3 && messageBuffer[messageBufferIndex] != magicBytes[messageBufferIndex])
-            {
-                memset(messageBuffer, 0, messageBufferIndex + 1);
-                messageBufferIndex = 0;
-            }
+        if(messageBufferIndex <= 3 && messageBuffer[messageBufferIndex] != magic_bytes[messageBufferIndex])
+        {
+            memset(messageBuffer, 0, messageBufferIndex + 1);
+            messageBufferIndex = 0;
+        }
 
-            else if(messageLengthArray[ messageBuffer[4] ] <= messageBufferIndex + 1)
-            {
-                pthread_mutex_lock(&state->sockfd_mutex);
-                printf("Writing %d bytes to sockdf\n", messageBufferIndex + 1);
-                write(state->sockfd, messageBuffer, messageBufferIndex + 1);
-                pthread_mutex_unlock(&state->sockfd_mutex);
+        else if(messageLengthArray[ messageBuffer[4] ] <= messageBufferIndex + 1)
```

Comment:
A few things:

- `messageBuffer[4] ` is a byte that potentially was just read in from the serial line. Do you really want to use it to be indexing an array? If it has any value other than 0 or 1, memory corruption will occur.
- `<= messageBufferIndex + 1` is usually just written as `< messageBufferIndex`
- so basically you are just saying that there are two fixed sizes that messages can be: 9 and 6? Why not just have the 4th byte of the message contain the actual length of the message? That seems like it would be much more intuitive than passing a 0 for a length of 9 and a 1 for a length of 6. That way you aren't dependent on some arbitrarily defined array of message lengths. The length of the message is directly encoded into the message itself.